language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -276,146 +351,92 @@ computeCek con t@(Var _ varName)   = do
-    term             ->
-        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just term
+    => Context uni -> CekVal uni -> CekM uni (Plain Term uni)
+--- Instantiate all the free variable of the resulting term in case there are any.
+-- . ◅ V           ↦  [] V
+returnCek [] val = pure $ void $ dischargeCekVal val
+-- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
+returnCek (FrameTyInstArg ty : ctx) fun = instantiateEvaluate ctx ty fun
+-- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
+    -- funVarEnv <- getEnv
+    withEnv argVarEnv $ computeCek (FrameApplyFun fun : ctx) arg
+-- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
+returnCek (FrameApplyFun fun : ctx) arg = do
+    applyEvaluate ctx fun arg
+-- s , wrap A B _ ◅ V  ↦  s ◅ wrap A B V
+returnCek (FrameIWrap ex pat arg : ctx) val =
+    returnCek ctx $ VIWrap ex pat arg val
+-- s , unwrap _ ◅ wrap A B V  ↦  s ◅ V
+returnCek (FrameUnwrap : ctx) val =
+    case val of
+      VIWrap _ _ _ v -> returnCek ctx v
+      _              ->
+        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just val
 
 -- | Instantiate a term with a type and proceed.
--- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
--- apply the term to the type via 'TyInst'.
+-- In case of 'VTyAbs' just ignore the type; for 'VBuiltin', extend
+-- the type arguments with the type, decrement the argument count,
+-- and proceed; otherwise, it's an error.
 instantiateEvaluate
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just fun
-
-{- Note [Saved mapping example]
-Consider a polymorphic built-in function @id@, whose type signature on the Plutus side is
-
-    id : all a. a -> a
-
-Notation:
-
-- the variable environment is denoted as @{ <var> :-> (<env>, <value>), ... }@
-- the context is denoted as a Haskell list of 'Frame's
-- 'computeCek' is denoted as @(|>)@
-- 'returnCek' is denoted as @(<|)@
-
-When evaluating
+    => Context uni -> Type TyName uni ExMemory -> CekVal uni -> CekM uni (Plain Term uni)
+instantiateEvaluate ctx _ (VTyAbs _ _ _ body env) = withEnv env $ computeCek ctx body -- FIXME: env?
+instantiateEvaluate ctx ty (VBuiltin ex argEnv bn count tyargs args) =
+    returnCek ctx $ VBuiltin ex argEnv bn (count-1) (ty:tyargs) args
+    -- We should really check that the signature expects a type here, not a term.
```

Comment:
I think that's okay: that's the typechecker's job.

Reply:
There's no type checker in the untyped world. We should check as many things during evaluation as we can, I believe. Even if it slows things down. At least that should be the default behaviour, it's good to provide some additional safety when you juggle lots of money.

Reply:
> We should check as many things during evaluation as we can

Maybe that's something that we should discuss.  My assumption has been that it's up to the participants in a transaction to check that the scripts do what they should, and one thing you might want to do there is typecheck a typed script off-chain and check that it erases to the untyped on-chain thing (and we still have the problem of knowing how your typed PLC came from some specific Haskell source).  We don't have a typechecker in the untyped world, but you shouldn't be agreeing to run stuff that you haven't typechecked beforehand.

Now I say all that, but it's very debatable.  Maybe the question of how defensive we should be during evaluation depends on some kind of social/moral policy about how we expect customers to behave and what their responsibilities are.  That's perhaps not an issue for us to decide (although since we're in blockchain world the answer will probably be that if you get scammed it's your own fault).

Reply:
> We don't have a typechecker in the untyped world, but you shouldn't be agreeing to run stuff that you haven't typechecked beforehand.

What if we somehow end up having a bug in the type checker? We do have one right now where we break global uniqueness in `unfoldFixOf`. It's better to double-check.

Moreover, the chain language is going to be Untyped Plutus Core and people may compile to it from something that is not Plutus at all and doesn't have types. We still should ensure that `force` is used consistently with `delay`/builtins, I believe.